### ISSUE TITLE
Quantization method for conv, deconv and fc layers.

### ISSUE THREAD

**Issue by Wangyida** [nyanp/tiny-cnn#202]
## Quantization Method for conv, deconv and fc Layers.
Here I want to implement the quanzization on operation in conv, deconv and fc layers. Much quantization method are included in this paper: Ristretto: Hardware-Oriented Approximation of Convolutional Neural [https://arxiv.org/pdf/1605.06402v1.pdf](url)

### Applicable Method
@nyanp Methods are mainly about 3 aspact:

1. **Network structure modification**: 
SqueezeNet [http://arxiv.org/pdf/1602.07360v3.pdf](url), ENet [https://arxiv.org/pdf/1606.02147v1.pdf](url)
2. **Binarization**: 
XNOR-Net [http://arxiv.org/pdf/1603.05279v2.pdf](url), Binary Connect [http://arxiv.org/pdf/1511.00363v3.pdf](url)
3. **Precision Adjustment**: 
Ristretto [http://arxiv.org/pdf/1606.01981v1.pdf](url)

### For tiny_cnn Implementaion
So there I propose some method for further implementation:

1. To utilize methods about network modification method, Inception concept could solve parameter expanding from large kernel scale (7 * 7, 9 * 9 and 11 * 11) and large amount of channels (AlexNet), This is related to the graph implementation discussed in #201 @edgarriba already give me some advises on the graph style.
2. This will bring an obvious accuracy drop, so I might not going to apply it.
3. @bhack This is the **main task** for me: Quantization to Fixed Point. I think it is better to implement a dynamic fix point data to replace float data, this could be done separately in conv, deconv and fc layers for the kernel file containing forward and backward codes.  It is desirable to group fixed point numbers into groups with constant  the fractional length,  so the number of bits allocated to the fractional part is constant within that group, but different compared to other groups. The fractional length learned to cover
a wide dynamic range of output and weight in particular layers. So there will be another worker storage or output channel to contain the fractional length.

**Comment by bhack** [comment]
We need also to see if we want to use https://github.com/google/gemmlowp kernles engine for 8bit  (or lower) as in TF.

**Comment by bhack** [comment]
@petewarden Do you plan to continue to support this kernels outside TF?

**Comment by petewarden** [comment]
I expect to continue with development of gemmlowp as a project in its own right, since it's a useful module for a lot of purposes.

**Comment by bhack** [comment]
@petewarden Thanks. We will try to rely on it. SSE4 is it still in the roadmap?

**Comment by Wangyida** [comment]
@bhack Good idea, Tensorflow applies gemmlowp kernels seemingly in quantization module [https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/quantization/kernels](url), its quantization module [https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/quantization](url) explain that the output and input still uses float data and the the MAC(multiplication and accumulation) operation is based on the quantified data which could be 8 bits. So the input and output in a single layer is quantized and dequantized using a desirable data type. Could I have a try on implement similar optimization?

 @petewarden Is it means that as for the MAC operation, [https://github.com/google/gemmlowp](url) is the core?

**Comment by bhack** [comment]
@Wangyida Yes could be interesting. I have not already seen how this quantization strategy compare to Ristretto [approximation schemes](http://lepsucd.com/?page_id=639)

**Comment by bhack** [comment]
@Wangyida A little bit was described by Pete in https://petewarden.com/2016/05/03/how-to-quantize-neural-networks-with-tensorflow/

**Comment by Wangyida** [comment]
@bhack I have read through the blog and it is useful indeed. Quantization and dequantization is needed because data between different layers should have a float type for a good performance, so the main operation within layers are just 2 tasks:
1. Quantization and dequantization using 8 bits data.
2. Having a decision whether quantization is needed in particular step.

tensorflow provided an example for the quantization method [https://www.tensorflow.org/versions/r0.9/how_tos/quantization/index.html](url) and I am running this one. I think the main difference between is that Ristretto try to seeking a better fix point (**dynamic fix point**) data to simulate float according to the range of the original float data and the method in [https://petewarden.com/2016/05/03/how-to-quantize-neural-networks-with-tensorflow/](url) is more likely an 8 bits representation for original float data and **quantization-dequantization** pair is needed for MAC operation. They both need to have the range of original float data which differs in different layer and different layer types described here [https://arxiv.org/pdf/1605.06402v1.pdf](url).

Maybe the method introduced in quantization module of tensorflow is more applicable compared to Ristretto which is developed in Caffe.

**Comment by Wangyida** [comment]
@bhack Could I open new PR about quantization and dequantization? This jointed operation on the input and output for particular layers could contribute to 4 times squeeze for network parameter. These steps used in the tensorflow module could be added outside the kernel functions in tiny_cnn in particular layer headers or standalone kernels names as 'quantization'. I had a read in the codes and demo in tensorflow. Or maybe it is better done after PR about nnpack got merged.

**Comment by bhack** [comment]
I don't want to create too many bootleneck between students. If you can try to open a new PR.

**Comment by Wangyida** [comment]
OK then I will try to have an implementation on this, parameter compression ratio against float will be roughly 25% and the performence won't be damaged much

**Comment by bhack** [comment]
Yes we can always quantize enet archs in a second iteration to gain more.

**Comment by nyanp** [comment]
@Wangyida 
Very attractive ideas. Do you have a plan to implement task1/2?

**Comment by Wangyida** [comment]
@nyanp 
Tasks like these are not related to modification in codes. It could be implemented using the 'graph' of tiny_cnn rather than 'stream' style. Both ENet and SqueezeNet performs well on this aspect, I used to have experiments using Caffe.

**Comment by nyanp** [comment]
@Wangyida 
I got it. I'll try to read these papers, thx :)

**Comment by Wangyida** [comment]
@nyanp OK, These methods are mainly based on multi-inputs layers and small kernels.

**Comment by edgarriba** [comment]
I have started with the benchmarks models (alexnet). My plan is to add it into 'tiny_cnn/examples/benchmarks' and maybe it's a good place to include the models you mention @Wangyida. I'll commit it during the afternoon.

**Comment by Wangyida** [comment]
@edgarriba Nice to hear that.

**Comment by bhack** [comment]
@nyanp 2. it is something different...

**Comment by Wangyida** [comment]
Task 2 might have a big drop in accuracy which may not be so acceptable, but the core functions are also efficient and interesting.

**Comment by bhack** [comment]
For lower resolution see also [DoReFa-Net](http://arxiv.org/abs/1606.06160)

**Comment by Wangyida** [comment]
@bhack Think you, really new.

**Comment by nyanp** [comment]
@edgarriba
I think we can implement basic lstm cell on the graph. We may need to modify network class a bit to handle BPTT training.

**Comment by Wangyida** [comment]
@bhack @edgarriba A new version of quantized convolution is pushed.

**Comment by bhack** [comment]
@nyanp Would you test the graph API with an initial rnn/lstm experiment?

**Comment by Maratyszcza** [comment]
@petewarden Any reason why gemmlowp doesn't compute convolution directly (i.e. intergrate im2col)?

@bhack @Wangyida I would consider implementing 8-bit fixed-point functions for inference if there is interest in using them. However, performance would be about the same as FP32.

**Comment by Wangyida** [comment]
@Maratyszcza As described in [Pete's blog for quantization](https://petewarden.com/2016/05/03/how-to-quantize-neural-networks-with-tensorflow/), math operations like convolution or matrix multiplication which produce 32-bit accumulated results from 8-bit inputs should be done under 32-bit because the accumulated results will surely have more than 8 bits of precision.
From my own point of view, the real advantage lies in the ```max(float)```, ```min(float)``` and ```uint8``` representation for concatenated layers in CNN. So I think maybe this process is better:

1. quantize input and weights to ```uint8``` and records the real range.
2. do math operation under ```int32```.
3. dequantize output from ```int32``` to ```float``` or from ```int32``` to ```uint8``` together with ```max(float)```, ```min(float)``` for the next layer.

**Comment by Maratyszcza** [comment]
@Wangyida 8-bit inputs and weights are multiplied to produce 16-bit results, which are then accumulated to 32-bit outputs. On most architectures there are better ways to do it than extending inputs to 32 bits. The 32-bit outputs are then reduce back to 8-bit activations. This way the whole network can operate with 8-bit fixed-point activations.

**Comment by Wangyida** [comment]
@Maratyszcza When 32-bit outputs are reduced back again to 8-bits, the real benefit from parameter squeezing could be got throughout the whole network. There should be 8-bit input rather than 32-bits in each layer for operation. My structure is not a whole 8-bits based network now, it might be considered as a task later.

**Comment by bhack** [comment]
@Maratyszcza Neon kernels are interesting for lowpower target (embedded/mobile) and gemmlowp has more complete coverage on Neon. NNpack it is still FMA-AVX2 only.

**Comment by bhack** [comment]
How we are syncing with gemmlowp upstream updates?

**Comment by Wangyida** [comment]
gemmlowp is an option for forward pass codes of fully connected kernels in tiny-cnn. If gemmlowp is not detected, it will use simple quantized operation instead.

**Comment by bhack** [comment]
I mean hown it is maintained updated. There are new kernels implemention optimized by Arm added upstream.

**Comment by Wangyida** [comment]
This is not included in gemmlowp in tiny-cnn, we can updates the gemmlowp codes based on commits [here](https://github.com/google/gemmlowp/commit/a00844b844482aae18c9bfd07bbea3e6e538ba3a)

**Comment by bhack** [comment]
The choice is if it makes sense to use git submodule or manually comply gemmlowp file every time

**Comment by Wangyida** [comment]
git submodule is good option cause most files from gemmlowp is included in tiny-cnn.

**Comment by edgarriba** [comment]
I think git submodule is not downloaded by default, a user has to explicitly call command `git submodule init`. So, in the end gemmlowp will be an optional lib. If so, more macros are needed in code.

**Comment by bhack** [comment]
Yes but manually monitoring git log on an not versioned library it is not easier.

**Comment by patrikhuber** [comment]
Is gemmlowp not header-only? I think git submodule for header-only is completely fine. You just have to add `--recursive` when cloning and everything else behaves and works the same as if the header was included directly. As a submodule it makes it much easier to maintain and update.

**Comment by edgarriba** [comment]
if the project provides tags is okay, otherwise you can expect that your API is broken at some point. It's true that with `--recursive` submodules are downloaded but it would be better to have the code with a simple `git clone`without additional flags since probably 90% of git users don't know or are not used with additional flags.

**Comment by patrikhuber** [comment]
Well, you can put it in one of the first lines of the readme.md. Whoever then doesn't read it will probably not be able to use the software anyway... :D In any case I agree it's one more (inconvenient) step, and the choice is based on personal opinion. I think it's outweighed by the fact that a user can immediately see that tiny-cnn uses an unmodified original version of gemmlowp from their original repo, and for tiny-cnn itself, it's reduced maintenance overhead.
 
I think it even works if the subproject doesn't use tags. By default, you can make it point to a specific commit, and it'll stay on that, so it won't break. Then one day when you want to update it, you can test your project with the git submodule pointing to the new commit id of the subproject, before you commit the change.

**Comment by bhack** [comment]
A new fix it is available upstream. See

**Comment by bhack** [comment]
Can we transform this to have a clear bullet point tasks list of what it is still missing?

**Comment by Wangyida** [comment]
@bhack An important fact which make the quantised operations not so efficient in code structure is that non-linear operations(activations) are treated as standalone operations in the codes structure. It will be much more tidy in quantised operations and do a head-to-tail quantised data transmission using "Quantised data - Range" pair if these operations could be transformed into operation kernels just like other linear kernels like conv, deconv and pool.
@nyanp @edgarriba So in my perspective, maybe standalone nonlinear operations are better than structure of tiny-dnn now. Moreover, the memory costs more when we apply padding in input and unpadding in back-propagated gradients which means that necessary data doesn't only exists in input/output, I think if there are some proper way to deal with data flows while the padding strategy is set same in convolution and deconvolution.

**Comment by bhack** [comment]
What is the status of https://github.com/tiny-dnn/tiny-dnn/pull/443? What else is needed?

**Comment by hanochk** [comment]
Can someone tells how QuantizedConv2D (tensorflow/core/kernels/quantized_conv_ops.cc) goes down to, Neon implementation in gemmlowp ? is it via QuantizedMatMul that utilises GemmlowpMultiply where  the singleThreadGemm receives the matrices and runs the compute() which implicitly calls the pack_neon, kernel and unpack?
